src/Microsoft.OData.Core/Json/JsonWriterAsync.cs

```diff
@@ -0,0 +1,448 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="JsonWriterAsync.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.Json
+{
+    #region Namespaces
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Globalization;
+    using System.IO;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.OData.Buffers;
+    using Microsoft.OData.Edm;
+    #endregion Namespaces
+
+    /// <summary>
+    /// Writer for the JSON format. http://www.json.org
+    /// </summary>
+    [SuppressMessage("Microsoft.Design", "CA1001:TypesThatOwnDisposableFieldsShouldBeDisposable", Justification = "This class does not own the underlying stream/writer and thus should never dispose it.")]
+    internal sealed partial class JsonWriter
+    {
+        /// <summary>
+        /// Start the padding function scope.
+        /// </summary>
+        public async Task StartPaddingFunctionScopeAsync()
+        {
+            Debug.Assert(this.scopes.Count == 0, "Padding scope can only be the outer most scope.");
+            await this.StartScopeAsync(ScopeType.Padding).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// End the padding function scope.
+        /// </summary>
+        public async Task EndPaddingFunctionScopeAsync()
+        {
+            Debug.Assert(this.scopes.Count > 0, "No scope to end.");
+
+            await this.writer.WriteLineAsync().ConfigureAwait(false);
+            await this.writer.DecreaseIndentationAsync().ConfigureAwait(false);

```

Comment:
Does the `DecreaseIndentationAsync()` actually write anything? Should it be async?

Reply:
@habbes Depending on the writer implementation, it might or might not write anything. For this reason, it's right to await. It just happens that for `NonIndentedTextWriter` it doesn't write anything.

Reply:
I understand that argument for `WriteLineAsync`, but I guess I just couldn't imagine an implementation of increasing/decreasing indentation that would need to write to the stream. I assume it would update the indentation level, which would probably be stored in an int, then other `Writexxx` methods would read the indentation level to decide how many spaces to add after a new line.

Reply:
@habbes On the contrary, since `WriteAsync`/`WriteValueAsync` methods do not concern themselves with whitespace characters, the typical implementation of `DecreaseIndentationAsync` would be responsible for working out the applicable indentation level while also writing the whitespace spaces required to apply the indentation.
For example, in `EndObjectScopeAsync` method, we do:
- `WriteLineAsync()` - Apply a new line
- `DecreaseIndentationAsync()` - Decrease indentation level and apply indentation
- `WriteAsync(scope.EndString)` - Write the `}` character to signal end of the object
From the above you can see that `DecreaseIndentationAsync` would have to be responsible for both decreasing indentation level and applying the indentation itself.

Reply:
Hmm I see what you mean. But here is where my confusion comes from.
`IncreaseIndentationAsync()` and `DecreaseIndentationAsync()` seem to only be called once per "block" (where a block is a series of "statements" that should appear at the same indentation level).

If the indentation method was responsible for writing the spaces, then it would be called every time before writing a property for example, or every time before writing an array element. Is that currently the case? Looking at the usage, it seems that the indentation methods are only called at the beginning and end of an object, and not at the properties in between. I could be wrong. And this could be irrelevant at the end of the day. But could be worth find out.

Reply:
Maybe it could make sense to measure whether making IncreaseIndentation async vs sync makes a difference. If it doesn't make a difference then I think you could ignore my concern. If it does make a difference, it could be worth looking into because once this method is merged in the public API, it might be hard to change later on if we find that we need to, since the signature would have to change.

Reply:
WebAPI JSON responses are indented, maybe we could see what implementation it uses, that could provide insights on how the Increase/DecreaseIndentation methods are meant to achieve. But if in doubt, I think making them async is the safer thing to do.

Reply:
Actually, just remembered that they are not indented, it's just that the postman client and browser extensions auto-format the response. My bad.

Reply:
@habbes 
> indentation methods are only called at the beginning and end of an object, and not at the properties in between
This is indeed the case. The indentation methods are called at the beginning of an array, object and padding function. The JsonWriter in ODL does not apply a newline or indentation before a property is written